runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -639,6 +639,165 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
       {
       switch (rm)
          {
+         case TR::java_lang_invoke_MethodHandle_asType:
+            {
+            TR::Node* mh = node->getChild(0);
+            TR::Node* mt = node->getChild(1);
+            bool mhConstraintGlobal, mtConstraintGlobal;
+            TR::VPConstraint* mhConstraint = getConstraint(mh, mhConstraintGlobal);
+            TR::VPConstraint* mtConstraint = getConstraint(mt, mtConstraintGlobal);
+            if (mhConstraint
+                && mtConstraint
+                && mhConstraint->getKnownObject()
+                && mhConstraint->isNonNullObject()
+                && mtConstraint->getKnownObject()
+                && mtConstraint->isNonNullObject())
+               {
+               uintptrj_t* mhLocation = getObjectLocationFromConstraint(mhConstraint);
+               uintptrj_t* mtLocation = getObjectLocationFromConstraint(mtConstraint);
+               uint32_t mtOffset = J9VMJAVALANGINVOKEMETHODHANDLE_TYPE_OFFSET(comp()->fej9()->vmThread());
+
+               TR::VMAccessCriticalSection asType(comp(),
+                                                           TR::VMAccessCriticalSection::tryToAcquireVMAccess);
+               if (!asType.hasVMAccess())
+                  break;
+               uintptrj_t mhObject = comp()->fej9()->getStaticReferenceFieldAtAddress((uintptrj_t)mhLocation);
+               uintptrj_t mtInMh = comp()->fej9()->getReferenceFieldAtAddress(mhObject + mtOffset);
```

Comment:
We should check before this access that `mhObject` really is an instance of `MethodHandle`. If we're (unknowingly) looking at a dead path, then it's possible to have some other type of object instead.